keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -0,0 +1,544 @@
+`PatchOptions`.
+
+Each apply call will be required to provide a fieldmanager name. We will not
+provide a a way for the fieldmanager name to be set for the entire
+clientset. There are a couple reasons for this:
+
+- If a client has multiple code paths where it makes apply requests to the same
+  object, but with different field sets, they must use different field manager
+  names. If they use the same field manager name they will cause fields to be
+  accidentally removed or disowned. This is a potential foot gun we would like to
+  avoid.
+
+- Apply requests always conflict with update requests, even if they were made by
+  the same client with the same field manager name. This is by design. So when a
+  controller migrates from update to apply, it will need to deal with conflicts
+  regardless of what field manager name is used.
+
+### Generated apply configuration types
+
+All fields present in an apply configuration become owned by the applier after
+when the apply request succeeds. Go structs contain zero valued fields which are
+included even if the user never explicitly sets the field. Required boolean
+fields are a good example of fields that would be applied incorrectly using go
+structs, e.g. `ContainerStatus.Ready` (required, not omitempty). Because of this
+we cannot use the existing go structs to represent apply configurations.
+
+<<[UNRESOLVED @jpbetz @jennybuckley ]>> 
+Finalize which alternative to use based on developer feedback. See the
+[Alternatives](#alternatives) for a complete list, but are currently focusing on
+the two below alternatives. We are working with the Kubebuilder community to
+gather feedback on what developers prefer.
+<<[/UNRESOLVED]>>
+
+#### Alternative 1: Genreated structs where all fields are pointers
+
+Example usage:
+
+```go
+&appsv1apply.Deployment{
+  Name: &pointer.StringPtr("nginx-deployment"),
+  Spec: &appsv1apply.DeploymentSpec{
+    Replicas: &pointer.Int32Ptr(0),
+    Template: &v1apply.PodTemplate{
+      Spec: &v1apply.PodSpec{
+        Containers: []v1.Containers{
+          {
+            Name: &pointer.StringPtr("nginx"),
+            Image: &pointer.StringPtr("nginx:latest"),
+          },
+        }
+      },
+    },
+  },
+}
+```
+
+For this approach, developers need to use a library like
+https://github.com/kubernetes/utils/blob/master/pointer/pointer.go
+to inline primitive literals.
+
+#### Alternative 2: Generated "builders"
+
+Example usage:
+
+```go
+&appsv1apply.Deployment().
+  SetObjectMeta(&metav1apply.ObjectMeta().
+    SetName("nginx-deployment")).
+  SetSpec(&appsv1apply.DeploymentSpec().
+    SetReplicas(0).
+    SetTemplate(
+      &v1apply.PodTemplate().
+        SetSpec(&v1apply.SetPodSpec().
+          SetContainers(v1apply.ContainerList{
+            v1apply.Container().
+              SetName("nginx").
+              SetImage("nginx:1.14.2")
+            v1apply.Container().
+              SetName("sidecar").
+          })
```

Comment:
This looks way better to me than option 1.

I think ideal would be embedding yaml directly, as in the tests for smd? But... that's not typesafe and there's no way to make it typesafe. Rust macros would be awesome for that. :(

Reply:
I expect this to be the most contentious part of this KEP 😬 

I started strongly in support of this option as well. I'm now on the fence because quite a few engineers I've shown the alternatives to strongly favored option 1. Their main reason seems to be a strong preference toward the use of go structs, even if they need to deal with lots of pointers to literals. 

Let's merge as provisional, gather the feedback from the controller-runtime/kubebuilder community and then discuss in a sig-apimachinery bi-weekly.

The idea of using macros in a language that supports it is really interesting.